controller/api/destination/endpoint_translator.go

```diff
@@ -491,33 +491,34 @@ func createWeightedAddr(address watcher.Address, opaquePorts map[uint32]struct{}
 	weightedAddr.MetricLabels = pkgK8s.GetPodLabels(address.OwnerKind, address.OwnerName, address.Pod)
 	_, isSkippedInboundPort := skippedInboundPorts[address.Port]
 
-	// If the pod is controlled by any Linkerd control plane, then it can be
-	// hinted that this destination knows H2 (and handles our orig-proto
-	// translation)
 	weightedAddr.ProtocolHint = &pb.ProtocolHint{}
-	if controllerNSLabel != "" && !isSkippedInboundPort {
+
+	_, opaquePort := opaquePorts[address.Port]
+	if address.OpaqueProtocol || opaquePort {
 		// If address is set as opaque by a Server, or its port is set as
-		// opaque by annotation or default value, then hint its proxy's
-		// inbound port, and set the hinted protocol to Opaque, so that the
-		// client proxy does not send a SessionProtocol.
-		_, opaquePort := opaquePorts[address.Port]
-		if address.OpaqueProtocol || opaquePort {
+		// opaque by annotation or default value, then set the hinted protocol to
+		// Opaque.
+		weightedAddr.ProtocolHint.Protocol = &pb.ProtocolHint_Opaque_{
+			Opaque: &pb.ProtocolHint_Opaque{},
+		}
+		if controllerNSLabel != "" && !isSkippedInboundPort {
 			port, err := getInboundPort(&address.Pod.Spec)
 			if err != nil {
 				log.Error(err)
+				return nil, err
```

Comment:
The error handling in this function is a bit all over the place:

It begins with returning an error without logging:
```go
	tcpAddr, err := toAddr(address)
	if err != nil {
		return nil, err
	}
```

then we have logging an error without returning:

```go
	skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(address.Pod)
	if err != nil {
		log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
	}
```

and now we have both:

```
			port, err := getInboundPort(&address.Pod.Spec)
			if err != nil {
				log.Error(err)
				return nil, err
			}
```

Is it really appropriate for each of these errors to be handled differently? My general expectation is that in most cases we should _either_ log _or_ return an error, but rarely should we do both.

Given that `createWeightedAddr` is a utility function that returns an error, my general expectation would be that the function wrap errors with, for example:

```
			port, err := getInboundPort(&address.Pod.Spec)
			if err != nil {
				return nil, fmt.Errorf("failed to read inbound port: %w", err)
			}
```

This means that we have to move the error log up to where the errors are being swallowed -- I assume before we return to the grpc server (or the nearest location we have sufficient logging context).